pkg/agent/openflow/pipeline_test.go

```diff
@@ -0,0 +1,237 @@
+					L2ForwardingOutTable,
+				},
+			},
+		},
+		{
+			ipStack: ipv4Only,
+			features: []feature{
+				&featurePodConnectivity{ipProtocols: ipStackMap[ipv4Only]},
+				&featureNetworkPolicy{enableAntreaPolicy: true},
+				&featureService{enableProxy: false},
+				&featureEgress{},
+			},
+			expectedTables: map[pipelineProtocol][]*FeatureTable{
+				pipelineIP: {
+					ClassifierTable,
+					SpoofGuardTable,
+					ConntrackTable,
+					ConntrackStateTable,
+					DNATTable,
+					AntreaPolicyEgressRuleTable,
+					EgressRuleTable,
+					EgressDefaultTable,
+					EgressMetricTable,
+					L3ForwardingTable,
+					L3DecTTLTable,
+					SNATTable,
+					L2ForwardingCalcTable,
+					AntreaPolicyIngressRuleTable,
+					IngressRuleTable,
+					IngressDefaultTable,
+					IngressMetricTable,
+					ConntrackCommitTable,
+					L2ForwardingOutTable,
+				},
+				pipelineARP: {
+					ARPSpoofGuardTable,
+					ARPResponderTable,
+				},
+			},
+		},
+		{
+			ipStack: ipv4Only,
+			features: []feature{
+				&featurePodConnectivity{ipProtocols: ipStackMap[ipv4Only]},
+				&featureNetworkPolicy{enableAntreaPolicy: true},
+				&featureService{enableProxy: true, proxyAll: false},
+				&featureEgress{},
+			},
+			expectedTables: map[pipelineProtocol][]*FeatureTable{
+				pipelineIP: {
+					ClassifierTable,
+					SpoofGuardTable,
+					SNATConntrackTable,
+					ConntrackTable,
+					ConntrackStateTable,
+					SessionAffinityTable,
+					ServiceLBTable,
+					EndpointDNATTable,
+					AntreaPolicyEgressRuleTable,
+					EgressRuleTable,
+					EgressDefaultTable,
+					EgressMetricTable,
+					L3ForwardingTable,
+					ServiceHairpinMarkTable,
+					L3DecTTLTable,
+					SNATTable,
+					SNATConntrackCommitTable,
+					L2ForwardingCalcTable,
+					AntreaPolicyIngressRuleTable,
+					IngressRuleTable,
+					IngressDefaultTable,
+					IngressMetricTable,
+					ConntrackCommitTable,
+					L2ForwardingOutTable,
+				},
+				pipelineARP: {
+					ARPSpoofGuardTable,
+					ARPResponderTable,
+				},
+			},
+		},
+	} {
+		templatesMap := make(map[pipelineProtocol][]*featureTemplate)
+		for _, f := range tc.features {
+			templatesMap[pipelineIP] = append(templatesMap[pipelineIP], f.getTemplate(pipelineIP))
+			if tc.ipStack != ipv6Only {
+				template := f.getTemplate(pipelineARP)
+				if template != nil {
+					templatesMap[pipelineARP] = append(templatesMap[pipelineARP], template)
+				}
+			}
+		}
+
+		for proto, templates := range templatesMap {
+			generatePipeline(templates)
+			tables := tc.expectedTables[proto]
+
+			for i := 0; i < len(tables)-1; i++ {
+				require.NotNil(t, tables[i].ofTable, fmt.Sprintf("table %s should be initialized", tables[i].name))
+				require.Less(t, tables[i].GetID(), tables[i+1].GetID(), fmt.Sprintf("id of table %s should less than that of table %s", tables[i].GetName(), tables[i+1].GetName()))
+			}
+			require.NotNil(t, tables[len(tables)-1].ofTable, fmt.Sprintf("table %s should be initialized", tables[len(tables)-1].name))
```

Comment:
there is no need to use `fmt.Sprintf` for error messages when using `assert` / `require`. check the documentation for the library,

Reply:
Thanks, updated.